File: server/migration/path.ts
import { exists, mkdir, readdir, rename, rmdir } from "node:fs/promises";
import { join } from "node:path";
import { publicDir } from "$/persist";

async function removeDirIfEmpty(dirPath: string) {
    try {
        const items = await readdir(dirPath, { withFileTypes: true });

        // Recurse into any nested directories first
        for (const item of items) {
            const fullPath = join(dirPath, item.name);
            if (item.isDirectory()) {
                // Go deeper
                await removeDirIfEmpty(fullPath);
            } else {
                // It contains a file, so this parent directory cannot be safely deleted
                console.log(`> Old directory is not empty. Skipping...`);
                return false;
            }
        }

        // Now that the subdirectories are cleaned out, try deleting this folder
        await rmdir(dirPath);
        console.log(`Successfully removed: ${dirPath}`);
        return true;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (error: any) {
        if (error.code === "ENOENT") {
            console.log(`> Old directory does not exist.`);
            return false;
        }

        throw error;
    }
}

async function moveFolderContents(sourceDir: string, targetDir: string) {
    try {
        // Check if the source directory exists
        if (!(await exists(sourceDir))) {
            return true;
        }

        console.log("> Migating public path...");

        // Ensure the target directory exists (creates it if not)
        await mkdir(targetDir, { recursive: true });

        // Read all files/folders from the source
        const files = await readdir(sourceDir);

        // Move each item
        for (const file of files) {
            const oldPath = join(sourceDir, file);
            const newPath = join(targetDir, file);

            await rename(oldPath, newPath);
        }

        // Remove the source directory if it's empty
        await removeDirIfEmpty(sourceDir);

        return true;
    } catch (error) {
        console.error("> An error occurred:", error);
        return false;
    }
}

export async function migratePublicPath() {
    if (process.env.NODE_ENV !== "production") return true;

    return moveFolderContents(publicDir("public"), publicDir());
}

File: server/migration/path.go
package migration

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"premark/persist"
)

// RenameAndCombine moves oldPath to newPath.
// It skips if oldPath doesn't exist, and merges contents if newPath already exists.
func RenameAndCombine(oldPath, newPath string) error {
	srcStat, err := os.Stat(oldPath)
	if os.IsNotExist(err) {
		// Old path doesn't exist
		return nil
	}
	if err != nil {
		return fmt.Errorf("Failed to stat source: %w", err)
	}
	if !srcStat.IsDir() {
		return errors.New("Source path is a file, not a directory")
	}

	dstStat, err := os.Stat(newPath)
	if os.IsNotExist(err) {
		// Destination doesn't exist at all, we can do a simple, fast rename
		return os.Rename(oldPath, newPath)
	}
	if err != nil {
		return fmt.Errorf("Failed to stat destination: %w", err)
	}
	if !dstStat.IsDir() {
		return errors.New("Destination path exists but is a file, cannot combine")
	}

	// Destination exists: time to recursively merge
	return mergeDirs(oldPath, newPath)
}

// mergeDirs recursively moves contents from src to dst, then deletes src
func mergeDirs(src, dst string) error {
	// Ensure the destination folder exists (with appropriate permissions)
	if err := os.MkdirAll(dst, 0755); err != nil {
		return err
	}

	entries, err := os.ReadDir(src)
	if err != nil {
		return err
	}

	for _, entry := range entries {
		srcItem := filepath.Join(src, entry.Name())
		dstItem := filepath.Join(dst, entry.Name())

		if entry.IsDir() {
			// If it's a sub-directory, recurse into it
			if err := mergeDirs(srcItem, dstItem); err != nil {
				return err
			}
		} else {
			// If it's a file, move it over (overwriting if necessary).
			if err := os.Rename(srcItem, dstItem); err != nil {
				return err
			}
		}
	}

	// Once all contents are safely moved out, remove the now-empty source directory
	return os.Remove(src)
}

func MigratePath() error {
	fmt.Println("> Migrating paths...")

	pathMap := map[string]string{
		persist.PublicDir("output", "presence_qr"): persist.PublicDir("output", "presence"),
	}

	for oldPath, newPath := range pathMap {
		if err := RenameAndCombine(oldPath, newPath); err != nil {
			return fmt.Errorf("Failed to migrate path %s to %s: %w", oldPath, newPath, err)
		}
	}

	return nil
}
